src/core/meshcop/joiner.cpp

```diff
@@ -0,0 +1,280 @@
+/*
+ *  Copyright (c) 2016, Nest Labs, Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Joiner role.
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <openthread-config.h>
+
+#include <common/code_utils.hpp>
+#include <common/encoding.hpp>
+#include <common/logging.hpp>
+#include <meshcop/joiner.hpp>
+#include <thread/thread_netif.hpp>
+#include <thread/thread_uris.hpp>
+
+using Thread::Encoding::BigEndian::HostSwap16;
+
+namespace Thread {
+namespace MeshCoP {
+
+Joiner::Joiner(ThreadNetif &aNetif):
+    mTransmitMessage(NULL),
+    mTransmitTask(&HandleUdpTransmit, this),
+    mJoinerEntrust(OPENTHREAD_URI_JOINER_ENTRUST, &HandleJoinerEntrust, this),
+    mNetif(aNetif)
+{
+    mNetif.GetCoapServer().AddResource(mJoinerEntrust);
+}
+
+ThreadError Joiner::Start(void)
+{
```

Comment:
How does Joiner begin to attach to the network after being entrusted?
Seems that we need to thread start manually?

Also for the on-mesh commissioner(not leader), when does it begin to attach after petition?

How does the user input PSKc, PSKd, Thread Network Credentials and etc. for commissioning?

How do we check the status of DTLS, commissioning status (both Joiner and Commissioner), and the petition status?